src/object/cli_obj.c

```diff
@@ -2375,6 +2407,21 @@ obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
 					shard_task_reset_param, &reset_arg);
 			}
 			goto task_sched;
+		} else if (require_shard_task) {
+			/*
+			 * The absence of shard tasks indicates that the epoch
+			 * was already chosen in the previous attempt. In this
+			 * attempt, since an epoch has not been chosen yet, the
+			 * TX must have been restarted between the two
+			 * attempts. This operation, therefore, is no longer
+			 * relevant for the restarted TX.
+			 *
+			 * This is only a temporary workaround; we will prevent
+			 * this case from happening in the first place, by
+			 * aborting and waiting for associated operations when
+			 * restarting a TX.
+			 */
+			return -DER_OP_CANCELED;
```

Comment:
just curious did you meet this case in test? how could it happen, an OP in a TX, retried and then find no epoch set for it...

Reply:
No, it would be quite difficult to trigger. Fan Yong and I came up with this case by reasoning. There could be a dc_tx_restart immediately before the retry.